File: 5mbps/producer.cc
#include <fstream>
#include <chrono>
#include <thread>
#include <librdkafka/rdkafkacpp.h>

#define TARGET_BPS 625000
#define HEADER_SIZE 24

int main() {
    std::ifstream f("data.ch10", std::ios::binary);
    std::string errstr;

    RdKafka::Conf *conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
    conf->set("bootstrap.servers", "localhost:9092", errstr);
    RdKafka::Producer *producer = RdKafka::Producer::create(conf, errstr);

    long sent = 0;
    auto start = std::chrono::steady_clock::now();

    while (f) {
        char sync[2];
        f.read(sync, 2);
        if (!f || sync[0] != 0x25 || sync[1] != (char)0xEB) continue;

        char header[HEADER_SIZE];
        header[0] = sync[0];
        header[1] = sync[1];
        f.read(header + 2, HEADER_SIZE - 2);

        uint32_t plen = *(uint32_t*)(header + 4);
        std::string packet(header, HEADER_SIZE);
        packet.resize(plen);
        f.read(&packet[HEADER_SIZE], plen - HEADER_SIZE);

        producer->produce(
            "ch10.raw.frames",
            RdKafka::Topic::PARTITION_UA,
            RdKafka::Producer::RK_MSG_COPY,
            packet.data(), packet.size(),
            nullptr, 0, 0, nullptr
        );

        sent += packet.size();
        auto now = std::chrono::steady_clock::now();
        double elapsed = std::chrono::duration<double>(now - start).count();
        if (elapsed > 0 && sent / elapsed > TARGET_BPS)
            std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }

    producer->flush(5000);
    delete producer;
    delete conf;
}

File: 23122025/producer.py
import struct
from kafka import KafkaProducer

CH10_SYNC = b'\x25\xEB'
HEADER_SIZE = 24

producer = KafkaProducer(
    bootstrap_servers='localhost:9092',
    value_serializer=lambda v: v
)

def produce_ch10_frames(ch10_file):
    frame_index = 1

    with open(ch10_file, "rb") as f:
        while True:
            b = f.read(1)
            if not b:
                break

            if b != CH10_SYNC[:1]:
                continue

            if f.read(1) != CH10_SYNC[1:]:
                f.seek(-1, 1)
                continue

            header = CH10_SYNC + f.read(HEADER_SIZE - 2)
            packet_len = struct.unpack("<I", header[4:8])[0]
            packet = header + f.read(packet_len - HEADER_SIZE)

            producer.send(
                topic="ch10.raw.frames",
                key=frame_index.to_bytes(4, "little"),
                value=packet
            )

            frame_index += 1

    producer.flush()
    print("DONE producing frames")

file_name = "../ch10_pcm_decoder/Calculex-PCM20_12052009_212115/data.ch10"

produce_ch10_frames(file_name)
